File: okay-compose/protocol/src/main/kotlin/okay/compose/protocol/Tree.kt
// What a client does with the tree: apply the server's patches
// (paths index children in order — docs/protocol/frontend.md), and
// the HYBRID rule: a Form's fields fold here, its button submits once,
// a live input speaks per change, everything else crosses the wire.
package okay.compose.protocol

object Tree {

    // ------------------------------------------------------------ patch

    fun apply(tree: Ui, p: Patch): Ui = when (p) {
        is Patch.Replace -> at(tree, p.path) { p.ui }
        is Patch.SetText -> at(tree, p.path) { u -> if (u is Ui.Text) u.copy(s = p.s) else u }
        is Patch.SetValue -> at(tree, p.path) { u -> if (u is Ui.Input) u.copy(value = p.s) else u }
        is Patch.SetChecked -> at(tree, p.path) { u -> if (u is Ui.Check) u.copy(on = p.on) else u }
        is Patch.SetSelected -> at(tree, p.path) { u -> if (u is Ui.Select) u.copy(selected = p.index) else u }
        is Patch.Remove -> at(tree, p.path) { u -> kids(u) { c -> c.filterIndexed { i, _ -> i != p.index } } }
        is Patch.Reorder -> at(tree, p.path) { u -> kids(u) { c -> p.order.map { c[it] } } }
        is Patch.Insert -> at(tree, p.path) { u -> kids(u) { c -> c.toMutableList().apply { add(p.index, p.ui) } } }
    }

    private fun at(u: Ui, path: List<Int>, f: (Ui) -> Ui): Ui {
        if (path.isEmpty()) return f(u)
        val i = path[0]
        val rest = path.drop(1)
        return when (u) {
            is Ui.Row -> u.copy(children = u.children.set(i) { at(it, rest, f) })
            is Ui.Column -> u.copy(children = u.children.set(i) { at(it, rest, f) })
            is Ui.Box -> u.copy(children = u.children.set(i) { at(it, rest, f) })
            is Ui.Scroll -> if (i == 0) u.copy(child = at(u.child, rest, f)) else u
            is Ui.Form -> u.copy(fields = u.fields.set(i) { at(it, rest, f) })
            is Ui.Items -> u.copy(items = u.items.set(i) { at(it, rest, f) })
            is Ui.Modal -> if (i == 1) u.copy(body = at(u.body, rest, f)) else u
            is Ui.Disclosure -> if (i == 1) u.copy(body = at(u.body, rest, f)) else u
            else -> u   // a path into a leaf: the server never makes one
        }
    }

    private fun kids(u: Ui, f: (List<Ui>) -> List<Ui>): Ui = when (u) {
        is Ui.Row -> u.copy(children = f(u.children))
        is Ui.Column -> u.copy(children = f(u.children))
        is Ui.Box -> u.copy(children = f(u.children))
        is Ui.Form -> u.copy(fields = f(u.fields))
        is Ui.Items -> u.copy(items = f(u.items))
        else -> u
    }

    private fun List<Ui>.set(i: Int, f: (Ui) -> Ui): List<Ui> =
        if (i < 0 || i >= size) this else mapIndexed { j, u -> if (j == i) f(u) else u }

    // ----------------------------------------------------------- hybrid

    /** every Form on the tree: its key, and its fields' keys */
    fun forms(u: Ui): Map<String, Set<String>> = when (u) {
        is Ui.Row -> u.children.flatMap { forms(it).entries }.associate { it.key to it.value }
        is Ui.Column -> u.children.flatMap { forms(it).entries }.associate { it.key to it.value }
        is Ui.Box -> u.children.flatMap { forms(it).entries }.associate { it.key to it.value }
        is Ui.Scroll -> forms(u.child)
        is Ui.Form -> u.fields.flatMap { forms(it).entries }.associate { it.key to it.value } +
            (u.key to u.fields.flatMap { keys(it) }.toSet())
        is Ui.Items -> u.items.flatMap { forms(it).entries }.associate { it.key to it.value }
        is Ui.Table -> u.rows.flatten().flatMap { forms(it).entries }.associate { it.key to it.value }
        is Ui.Tabs -> u.pages.getOrNull(u.selected)?.let { forms(it) } ?: emptyMap()
        is Ui.Modal -> forms(u.body)
        is Ui.Disclosure -> if (u.open) forms(u.body) else emptyMap()
        else -> emptyMap()
    }

    /** the capability list: every key an event may name */
    fun keys(u: Ui): Set<String> = when (u) {
        is Ui.Row -> u.children.flatMap { keys(it) }.toSet()
        is Ui.Column -> u.children.flatMap { keys(it) }.toSet()
        is Ui.Box -> u.children.flatMap { keys(it) }.toSet()
        is Ui.Scroll -> keys(u.child)
        is Ui.Text, is Ui.Image -> emptySet()
        is Ui.Button -> setOf(u.key)
        is Ui.Input -> setOf(u.key)
        is Ui.Check -> setOf(u.key)
        is Ui.Select -> setOf(u.key)
        is Ui.Form -> u.fields.flatMap { keys(it) }.toSet() + u.key
        is Ui.Items -> u.items.flatMap { keys(it) }.toSet()
        is Ui.Table -> u.rows.flatten().flatMap { keys(it) }.toSet()
        is Ui.Tabs -> u.labels.indices.map { "${u.key}\$tab$it" }.toSet() + (u.pages.getOrNull(u.selected)?.let { keys(it) } ?: emptySet())
        is Ui.Modal -> keys(u.body)
        is Ui.Disclosure -> (if (u.open) keys(u.body) else emptySet()) + u.key
    }

    /** the widgets in tab order — the interactive leaves */
    fun focusable(u: Ui): List<Ui> = when (u) {
        is Ui.Row -> u.children.flatMap { focusable(it) }
        is Ui.Column -> u.children.flatMap { focusable(it) }
        is Ui.Box -> u.children.flatMap { focusable(it) }
        is Ui.Scroll -> focusable(u.child)
        is Ui.Text, is Ui.Image -> emptyList()
        is Ui.Button, is Ui.Input, is Ui.Check, is Ui.Select -> listOf(u)
        is Ui.Form -> u.fields.flatMap { focusable(it) } + Ui.Button(u.submit, u.key, Role.Primary)
        is Ui.Items -> u.items.flatMap { focusable(it) }
        is Ui.Table -> u.rows.flatten().flatMap { focusable(it) }
        is Ui.Tabs -> u.labels.mapIndexed { i, l -> Ui.Button(l, "${u.key}\$tab$i", if (i == u.selected) Role.Active else Role.Plain) } +
            (u.pages.getOrNull(u.selected)?.let { focusable(it) } ?: emptyList())
        is Ui.Modal -> focusable(u.body)
        is Ui.Disclosure -> listOf(Ui.Button(u.title, u.key, if (u.open) Role.Active else Role.Plain)) + (if (u.open) focusable(u.body) else emptyList())
    }

    /** a bottom-up rewrite of every node */
    fun map(u: Ui, f: (Ui) -> Ui): Ui = f(
        when (u) {
            is Ui.Row -> u.copy(children = u.children.map { map(it, f) })
            is Ui.Column -> u.copy(children = u.children.map { map(it, f) })
            is Ui.Box -> u.copy(children = u.children.map { map(it, f) })
            is Ui.Scroll -> u.copy(child = map(u.child, f))
            is Ui.Form -> u.copy(fields = u.fields.map { map(it, f) })
            is Ui.Items -> u.copy(items = u.items.map { map(it, f) })
            is Ui.Table -> u.copy(rows = u.rows.map { r -> r.map { map(it, f) } })
            is Ui.Tabs -> u.copy(pages = u.pages.map { map(it, f) })
            is Ui.Modal -> u.copy(body = map(u.body, f))
            is Ui.Disclosure -> u.copy(body = map(u.body, f))
            else -> u
        },
    )

    /**
     * The hybrid rule, on the client: an event that stays local answers
     * the tree it changes; null means "send it". This client claims no
     * semantic node, so only the Form's fields are local here.
     */
    fun foldLocal(tree: Ui, e: Event): Ui? {
        val fieldOf = forms(tree).flatMap { (f, ks) -> ks.map { it to f } }.toMap()
        fun live(k: String) = focusable(tree).any { it is Ui.Input && it.key == k && it.live }
        return when (e) {
            is Event.Edited -> if (fieldOf.containsKey(e.key) && !live(e.key))
                map(tree) { u -> if (u is Ui.Input && u.key == e.key) u.copy(value = e.value) else u } else null
            is Event.Toggled -> if (fieldOf.containsKey(e.key))
                map(tree) { u -> if (u is Ui.Check && u.key == e.key) u.copy(on = e.on) else u } else null
            is Event.Chosen -> if (fieldOf.containsKey(e.key))
                map(tree) { u -> if (u is Ui.Select && u.key == e.key) u.copy(selected = e.index) else u } else null
            else -> null
        }
    }

    /** the Form's one event, from the values its fields hold now */
    fun submit(tree: Ui, formKey: String): Event? {
        var found: Ui.Form? = null
        map(tree) { u -> if (u is Ui.Form && u.key == formKey && found == null) found = u; u }
        val f = found ?: return null
        return Event.Submitted(
            formKey,
            f.fields.flatMap { focusable(it) }.mapNotNull {
                when (it) {
                    is Ui.Input -> Event.Edited(it.key, it.value)
                    is Ui.Check -> Event.Toggled(it.key, it.on)
                    is Ui.Select -> Event.Chosen(it.key, it.selected)
                    else -> null
                }
            },
        )
    }

    /** what the client sends for a host event: the local fold, the
     * Form's submit, or the event itself */
    fun outbound(tree: Ui, e: Event): Pair<Ui, Event?> {
        foldLocal(tree, e)?.let { return it to null }
        if (e is Event.Pressed && forms(tree).containsKey(e.key)) return tree to submit(tree, e.key)
        return tree to e
    }
}

File: okay-swift/Sources/OkayProtocol/Tree.swift
// What a client does with the tree: apply the server's patches (paths
// index children in order — docs/protocol/frontend.md), and the HYBRID
// rule: a Form's fields fold here, its button submits once, a live
// input speaks per change, everything else crosses the wire.
//
// On the Scala side each of these has a name in the optics
// (specs/optics.md stage 2), and porting one is easier knowing which:
//   apply(_:_:)  ~ Ui.path(patch.path)  — the affine at an index path
//   map(_:_:)    ~ Ui.everywhere        — the traversal over every node
//   keys / forms ~ Ui.shown             — only what is on screen
//   the key work ~ Ui.key(k)            — every node a key names
// A port needs none of that machinery; the names are the map.

public enum Tree {

    // ------------------------------------------------------------ patch

    public static func apply(_ tree: Ui, _ p: Patch) -> Ui {
        switch p {
        case let .replace(path, u): return at(tree, path) { _ in u }
        case let .setText(path, s): return at(tree, path) { u in if case let .text(_, st) = u { return .text(s: s, style: st) }; return u }
        case let .setValue(path, s): return at(tree, path) { u in
            if case let .input(_, k, l, kind, live) = u { return .input(value: s, key: k, label: l, kind: kind, live: live) }; return u }
        case let .setChecked(path, on): return at(tree, path) { u in if case let .check(_, k, l) = u { return .check(on: on, key: k, label: l) }; return u }
        case let .setSelected(path, i): return at(tree, path) { u in if case let .select(o, _, k) = u { return .select(options: o, selected: i, key: k) }; return u }
        case let .remove(path, i): return at(tree, path) { u in kids(u) { c in c.enumerated().filter { $0.offset != i }.map { $0.element } } }
        case let .reorder(path, order): return at(tree, path) { u in kids(u) { c in order.map { c[$0] } } }
        case let .insert(path, i, new): return at(tree, path) { u in kids(u) { c in var c = c; c.insert(new, at: i); return c } }
        }
    }

    static func at(_ u: Ui, _ path: [Int], _ f: (Ui) -> Ui) -> Ui {
        guard let i = path.first else { return f(u) }
        let rest = Array(path.dropFirst())
        switch u {
        case let .row(c, k): return .row(children: set(c, i) { at($0, rest, f) }, key: k)
        case let .column(c, k): return .column(children: set(c, i) { at($0, rest, f) }, key: k)
        case let .box(c, d, w, g, p, k): return .box(children: set(c, i) { at($0, rest, f) }, dir: d, weights: w, gap: g, pad: p, key: k)
        case let .scroll(c, k): return i == 0 ? .scroll(child: at(c, rest, f), key: k) : u
        case let .form(fs, s, k): return .form(fields: set(fs, i) { at($0, rest, f) }, submit: s, key: k)
        case let .items(it, k): return .items(items: set(it, i) { at($0, rest, f) }, key: k)
        case let .modal(t, b, k): return i == 1 ? .modal(title: t, body: at(b, rest, f), key: k) : u
        case let .disclosure(t, o, b, k): return i == 1 ? .disclosure(title: t, open: o, body: at(b, rest, f), key: k) : u
        default: return u   // a path into a leaf: the server never makes one
        }
    }

    static func kids(_ u: Ui, _ f: ([Ui]) -> [Ui]) -> Ui {
        switch u {
        case let .row(c, k): return .row(children: f(c), key: k)
        case let .column(c, k): return .column(children: f(c), key: k)
        case let .box(c, d, w, g, p, k): return .box(children: f(c), dir: d, weights: w, gap: g, pad: p, key: k)
        case let .form(fs, s, k): return .form(fields: f(fs), submit: s, key: k)
        case let .items(it, k): return .items(items: f(it), key: k)
        default: return u
        }
    }

    static func set(_ xs: [Ui], _ i: Int, _ f: (Ui) -> Ui) -> [Ui] {
        guard i >= 0 && i < xs.count else { return xs }
        var out = xs; out[i] = f(xs[i]); return out
    }

    // ----------------------------------------------------------- hybrid

    /// every Form on the tree: its key, and its fields' keys
    public static func forms(_ u: Ui) -> [String: Set<String>] {
        func merge(_ xs: [Ui]) -> [String: Set<String>] { xs.reduce(into: [:]) { $0.merge(forms($1)) { a, _ in a } } }
        switch u {
        case let .row(c, _), let .column(c, _), let .box(c, _, _, _, _, _): return merge(c)
        case let .scroll(c, _): return forms(c)
        case let .form(fs, _, k):
            var m = merge(fs); m[k] = Set(fs.flatMap { keys($0) }); return m
        case let .items(i, _): return merge(i)
        case let .table(_, rows, _): return merge(rows.flatMap { $0 })
        case let .tabs(_, s, pages, _): return s < pages.count ? forms(pages[s]) : [:]
        case let .modal(_, b, _): return forms(b)
        case let .disclosure(_, o, b, _): return o ? forms(b) : [:]
        default: return [:]
        }
    }

    /// the capability list: every key an event may name
    public static func keys(_ u: Ui) -> Set<String> {
        switch u {
        case let .row(c, _), let .column(c, _), let .box(c, _, _, _, _, _): return Set(c.flatMap { keys($0) })
        case let .scroll(c, _): return keys(c)
        case .text, .image: return []
        case let .button(_, k, _), let .input(_, k, _, _, _), let .check(_, k, _), let .select(_, _, k): return [k]
        case let .form(fs, _, k): return Set(fs.flatMap { keys($0) }).union([k])
        case let .items(i, _): return Set(i.flatMap { keys($0) })
        case let .table(_, rows, _): return Set(rows.flatMap { $0 }.flatMap { keys($0) })
        case let .tabs(l, s, pages, k):
            let tabs = Set(l.indices.map { "\(k)$tab\($0)" })
            return s < pages.count ? tabs.union(keys(pages[s])) : tabs
        case let .modal(_, b, _): return keys(b)
        case let .disclosure(_, o, b, k): return (o ? keys(b) : []).union([k])
        }
    }

    /// the interactive leaves, in order
    public static func focusable(_ u: Ui) -> [Ui] {
        switch u {
        case let .row(c, _), let .column(c, _), let .box(c, _, _, _, _, _): return c.flatMap { focusable($0) }
        case let .scroll(c, _): return focusable(c)
        case .text, .image: return []
        case .button, .input, .check, .select: return [u]
        case let .form(fs, s, k): return fs.flatMap { focusable($0) } + [.button(label: s, key: k, role: .primary)]
        case let .items(i, _): return i.flatMap { focusable($0) }
        case let .table(_, rows, _): return rows.flatMap { $0 }.flatMap { focusable($0) }
        case let .tabs(l, s, pages, k):
            let bar = l.enumerated().map { Ui.button(label: $0.element, key: "\(k)$tab\($0.offset)", role: $0.offset == s ? .active : .plain) }
            return bar + (s < pages.count ? focusable(pages[s]) : [])
        case let .modal(_, b, _): return focusable(b)
        case let .disclosure(t, o, b, k): return [.button(label: t, key: k, role: o ? .active : .plain)] + (o ? focusable(b) : [])
        }
    }

    /// a bottom-up rewrite of every node
    public static func map(_ u: Ui, _ f: (Ui) -> Ui) -> Ui {
        let inner: Ui
        switch u {
        case let .row(c, k): inner = .row(children: c.map { map($0, f) }, key: k)
        case let .column(c, k): inner = .column(children: c.map { map($0, f) }, key: k)
        case let .box(c, d, w, g, p, k): inner = .box(children: c.map { map($0, f) }, dir: d, weights: w, gap: g, pad: p, key: k)
        case let .scroll(c, k): inner = .scroll(child: map(c, f), key: k)
        case let .form(fs, s, k): inner = .form(fields: fs.map { map($0, f) }, submit: s, key: k)
        case let .items(i, k): inner = .items(items: i.map { map($0, f) }, key: k)
        case let .table(h, rows, k): inner = .table(header: h, rows: rows.map { $0.map { map($0, f) } }, key: k)
        case let .tabs(l, s, p, k): inner = .tabs(labels: l, selected: s, pages: p.map { map($0, f) }, key: k)
        case let .modal(t, b, k): inner = .modal(title: t, body: map(b, f), key: k)
        case let .disclosure(t, o, b, k): inner = .disclosure(title: t, open: o, body: map(b, f), key: k)
        default: inner = u
        }
        return f(inner)
    }

    /// the hybrid rule, on the client: an event that stays local answers
    /// the tree it changes; nil means "send it"
    public static func foldLocal(_ tree: Ui, _ e: Event) -> Ui? {
        let fieldOf = forms(tree).reduce(into: [String: String]()) { acc, kv in kv.value.forEach { acc[$0] = kv.key } }
        func live(_ k: String) -> Bool { focusable(tree).contains { if case let .input(_, key, _, _, l) = $0 { return key == k && l }; return false } }
        switch e {
        case let .edited(k, v):
            guard fieldOf[k] != nil, !live(k) else { return nil }
            return map(tree) { u in if case let .input(_, key, l, kind, lv) = u, key == k { return .input(value: v, key: key, label: l, kind: kind, live: lv) }; return u }
        case let .toggled(k, on):
            guard fieldOf[k] != nil else { return nil }
            return map(tree) { u in if case let .check(_, key, l) = u, key == k { return .check(on: on, key: key, label: l) }; return u }
        case let .chosen(k, i):
            guard fieldOf[k] != nil else { return nil }
            return map(tree) { u in if case let .select(o, _, key) = u, key == k { return .select(options: o, selected: i, key: key) }; return u }
        default: return nil
        }
    }

    /// the Form's one event, from the values its fields hold now
    public static func submit(_ tree: Ui, _ formKey: String) -> Event? {
        var found: [Ui]? = nil
        _ = map(tree) { u in if case let .form(fs, _, k) = u, k == formKey, found == nil { found = fs }; return u }
        guard let fields = found else { return nil }
        let edits: [Event] = fields.flatMap { focusable($0) }.compactMap {
            switch $0 {
            case let .input(v, k, _, _, _): return .edited(key: k, value: v)
            case let .check(on, k, _): return .toggled(key: k, on: on)
            case let .select(_, i, k): return .chosen(key: k, index: i)
            default: return nil
            }
        }
        return .submitted(key: formKey, edits: edits)
    }

    /// what the client sends for a host event: the local fold, the Form's submit, or the event itself
    public static func outbound(_ tree: Ui, _ e: Event) -> (Ui, Event?) {
        if let t = foldLocal(tree, e) { return (t, nil) }
        if case let .pressed(k) = e, forms(tree)[k] != nil { return (tree, submit(tree, k)) }
        return (tree, e)
    }
}
